File: iosApp/iosApp/Presentation/Main/CafeManagement/CafeManagementViewModel.swift
//
//  CafeManagementViewModel.swift
//  ConCafe
//
//  Created by 홍희표 on 2026/03/07.
//

import Foundation
import Combine
import Shared
import KMPNativeCoroutinesAsync

@MainActor
final class CafeManagementViewModel: ObservableObject {
    private let createCafeOwnerClaimUseCase: CreateCafeOwnerClaimUseCase

    private let getCafeManagementUseCase: GetCafeManagementUseCase

    private let observeCurrentUserUseCase: ObserveCurrentUserUseCase

    private let cafeRegistrationClaimEventPublisher: CafeRegistrationClaimEventPublisher

    private let cafeDetailEventPublisher: CafeDetailEventPublisher

    @Published private(set) var uiState = CafeManagementUiState()

    let event = PassthroughSubject<CafeManagementEvent, Never>()

    private var tasks: [TaskKey: Task<Void, Never>] = [:]

    private func loadCafeManagement() {
        tasks[.loadCafeManagement]?.cancel()
        tasks[.loadCafeManagement] = Task {
            uiState.isLoading = true
            do {
                let result = try await getCafeManagementUseCase.invoke()

                if let success = result as? AppResultSuccess<AnyObject>,
                   let data = success.data as? CafeManagementData {
                    uiState.ownedCafes = data.ownedCafes
                    uiState.searchableCafes = data.searchableCafes
                    uiState.pendingClaims = data.pendingClaims
                    uiState.isLoading = false
                } else if let failure = result as? AppResultFailure {
                    uiState.ownedCafes = []
                    uiState.searchableCafes = []
                    uiState.pendingClaims = []
                    uiState.isLoading = false
                    uiState.infoMessage = "\(failure.error)"
                }
            } catch {
                uiState.ownedCafes = []
                uiState.searchableCafes = []
                uiState.pendingClaims = []
                uiState.isLoading = false
                uiState.infoMessage = error.localizedDescription
            }
        }
    }

    private func refreshPendingClaims(resetMessage: Bool = true) {
        Task {
            do {
                let result = try await getCafeManagementUseCase.invoke()

                if let success = result as? AppResultSuccess<AnyObject>,
                   let data = success.data as? CafeManagementData {
                    uiState.ownedCafes = data.ownedCafes
                    uiState.searchableCafes = data.searchableCafes
                    uiState.pendingClaims = data.pendingClaims
                    if resetMessage {
                        uiState.infoMessage = nil
                    }
                } else if let failure = result as? AppResultFailure {
                    if resetMessage {
                        uiState.infoMessage = "\(failure.error)"
                    }
                }
            } catch {
                if resetMessage {
                    uiState.infoMessage = error.localizedDescription
                }
            }
        }
    }

    private func clickCafe(_ cafeId: String) {
        event.send(.navigateToCafeDashboard(cafeId: cafeId))
    }

    private func clickCafeDetail(_ cafeId: String) {
        event.send(.navigateToCafe(cafeId: cafeId))
    }

    private func changeCafeSearchQuery(_ query: String) {
        uiState.cafeSearchQuery = query
        uiState.infoMessage = nil
    }

    private func clickClaimCafe(_ cafeId: String) {
        Task {
            do {
                let result = try await createCafeOwnerClaimUseCase.invoke(cafeId: cafeId)

                if result is AppResultSuccess<AnyObject> {
                    refreshPendingClaims(resetMessage: false)
                    uiState.infoMessage = MessageKey.ownerClaimRegistered
                } else if let failure = result as? AppResultFailure {
                    uiState.infoMessage = "\(failure.error)"
                }
            } catch {
                uiState.infoMessage = error.localizedDescription
            }
        }
    }

    private func toggleCafeListExpanded() {
        uiState.isShowingAllCafes.toggle()
    }

    private func clickCreateCafe() {
        event.send(.navigateToCafeInfoRegistration)
    }

    private func dismissInfoMessage() {
        uiState.infoMessage = nil
    }

    private func observeSession() {
        tasks[.session]?.cancel()
        tasks[.session] = Task {
            do {
                for try await _ in asyncSequence(for: observeCurrentUserUseCase.invoke()) {
                    self.loadCafeManagement()
                }
            } catch {
                print("Error: \(error)")
            }
        }
    }

    private func observeCafeDetailEvent() {
        tasks[.cafeDetailEvent]?.cancel()
        tasks[.cafeDetailEvent] = Task {
            do {
                for try await event in asyncSequence(for: cafeDetailEventPublisher.events) {
                    if let updated = event as? CafeDetailEvent.CafeInfoUpdated {
                        self.patchCafeInfo(updated.cafe)
                    }
                }
            } catch {
                print("Error: \(error)")
            }
        }
    }

    private func observeCafeRegistrationClaimEvent() {
        tasks[.cafeRegistrationClaimEvent]?.cancel()
        tasks[.cafeRegistrationClaimEvent] = Task {
            do {
                for try await event in asyncSequence(for: cafeRegistrationClaimEventPublisher.events) {
                    let shouldRefresh: Bool
                    switch event {
                    case let created as CafeRegistrationClaimEvent.Created:
                        shouldRefresh = self.uiState.pendingClaims.contains(where: { pendingClaim in
                            pendingClaim.claimId == created.claimId
                        })
                    case let approved as CafeRegistrationClaimEvent.Approved:
                        shouldRefresh = self.uiState.pendingClaims.contains(where: { pendingClaim in
                            pendingClaim.claimId == approved.claimId
                        })
                    case let rejected as CafeRegistrationClaimEvent.Rejected:
                        shouldRefresh = self.uiState.pendingClaims.contains(where: { pendingClaim in
                            pendingClaim.claimId == rejected.claimId
                        })
                    default:
                        shouldRefresh = false
                    }
                    if shouldRefresh {
                        self.refreshPendingClaims(resetMessage: false)
                    }
                }
            } catch {
                print("Error: \(error)")
            }
        }
    }

    private func startClaimPolling() {
        let pollingIntervalNanoseconds = cafeManagementClaimPollingIntervalNanoseconds
        tasks[.claimPolling]?.cancel()
        tasks[.claimPolling] = Task { [weak self] in
            while !Task.isCancelled {
                do {
                    try await Task.sleep(nanoseconds: pollingIntervalNanoseconds)
                } catch {
                    break
                }

                if Task.isCancelled {
                    break
                } else {
                    self?.refreshPendingClaims(resetMessage: false)
                }
            }
        }
    }

    private func patchCafeInfo(_ cafe: Cafe) {
        uiState.ownedCafes = uiState.ownedCafes.map { item in
            guard item.id == cafe.id else { return item }
            return CafeManagementData.OwnedCafeSummary(
                id: item.id,
                name: cafe.name,
                city: cafe.region.city,
                isApproved: item.isApproved,
                todayVisitors: item.todayVisitors,
                todayCheckIns: item.todayCheckIns,
                todayReviews: item.todayReviews,
                rating: cafe.ratingAvg,
                castCount: item.castCount,
                noticeCount: item.noticeCount,
                externalLinkCount: item.externalLinkCount,
                thumbnailImage: cafe.thumbnailImage
            )
        }
        uiState.searchableCafes = uiState.searchableCafes.map { item in
            guard item.id == cafe.id else { return item }
            return CafeManagementData.SearchableCafeSummary(
                id: item.id,
                name: cafe.name,
                location: "\(cafe.region.city) \(cafe.region.address)"
            )
        }
    }

    func onAction(_ action: CafeManagementAction) {
        switch action {
        case .clickCafe(let cafeId):
            clickCafe(cafeId)
        case .clickCafeDetail(let cafeId):
            clickCafeDetail(cafeId)
        case .changeCafeSearchQuery(let query):
            changeCafeSearchQuery(query)
        case .clickClaimCafe(let cafeId):
            clickClaimCafe(cafeId)
        case .toggleCafeListExpanded:
            toggleCafeListExpanded()
        case .clickCreateCafe:
            clickCreateCafe()
        case .dismissInfoMessage:
            dismissInfoMessage()
        }
    }

    init(
        createCafeOwnerClaimUseCase: CreateCafeOwnerClaimUseCase = KoinInitializerKt.resolveCreateCafeOwnerClaimUseCase(),
        getCafeManagementUseCase: GetCafeManagementUseCase = KoinInitializerKt.resolveGetCafeManagementUseCase(),
        observeCurrentUserUseCase: ObserveCurrentUserUseCase = KoinInitializerKt.resolveObserveCurrentUserUseCase(),
        cafeRegistrationClaimEventPublisher: CafeRegistrationClaimEventPublisher = KoinInitializerKt.resolveCafeRegistrationClaimEventPublisher(),
        cafeDetailEventPublisher: CafeDetailEventPublisher = KoinInitializerKt.resolveCafeDetailEventPublisher()
    ) {
        self.createCafeOwnerClaimUseCase = createCafeOwnerClaimUseCase
        self.getCafeManagementUseCase = getCafeManagementUseCase
        self.observeCurrentUserUseCase = observeCurrentUserUseCase
        self.cafeRegistrationClaimEventPublisher = cafeRegistrationClaimEventPublisher
        self.cafeDetailEventPublisher = cafeDetailEventPublisher

        observeSession()
        observeCafeDetailEvent()
        observeCafeRegistrationClaimEvent()
        startClaimPolling()
    }

    deinit {
        tasks.values.forEach { $0.cancel() }
        tasks.removeAll()
    }

    private enum TaskKey {
        case loadCafeManagement
        case session
        case cafeDetailEvent
        case cafeRegistrationClaimEvent
        case claimPolling
    }

    private enum MessageKey {
        static let ownerClaimRegistered = "cafemgmt_info_owner_claim_registered"
    }
}

private let cafeManagementClaimPollingIntervalNanoseconds: UInt64 = 60_000_000_000

File: composeApp/src/commonMain/kotlin/com/hhp227/concafe/presentation/main/cafemanagement/CafeManagementViewModel.kt
package com.hhp227.concafe.presentation.main.cafemanagement

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import com.hhp227.concafe.domain.common.AppResult
import com.hhp227.concafe.domain.model.Cafe
import com.hhp227.concafe.domain.usecase.CreateCafeOwnerClaimUseCase
import com.hhp227.concafe.domain.event.CafeDetailEvent
import com.hhp227.concafe.domain.event.CafeRegistrationClaimEvent
import com.hhp227.concafe.domain.event.publisher.CafeDetailEventPublisher
import com.hhp227.concafe.domain.event.publisher.CafeRegistrationClaimEventPublisher
import com.hhp227.concafe.domain.usecase.GetCafeManagementUseCase
import com.hhp227.concafe.domain.usecase.ObserveCurrentUserUseCase

class CafeManagementViewModel(
    private val createCafeOwnerClaimUseCase: CreateCafeOwnerClaimUseCase,
    private val getCafeManagementUseCase: GetCafeManagementUseCase,
    private val observeCurrentUserUseCase: ObserveCurrentUserUseCase,
    private val cafeRegistrationClaimEventPublisher: CafeRegistrationClaimEventPublisher,
    private val cafeDetailEventPublisher: CafeDetailEventPublisher
) : ViewModel() {
    private val _uiState = MutableStateFlow(CafeManagementUiState())
    val uiState = _uiState.asStateFlow()

    private val _event = MutableSharedFlow<CafeManagementEvent>(replay = 0)
    val event = _event.asSharedFlow()

    private val jobs = mutableMapOf<TaskKey, Job>()

    private fun loadCafeManagement() {
        jobs[TaskKey.LOAD_CAFE_MANAGEMENT]?.cancel()
        jobs[TaskKey.LOAD_CAFE_MANAGEMENT] = viewModelScope.launch {
            _uiState.update { state ->
                state.copy(
                    isLoading = true
                )
            }
            when (val result = getCafeManagementUseCase.invoke()) {
                is AppResult.Success -> {
                    _uiState.update { state ->
                        state.copy(
                            ownedCafes = result.data.ownedCafes,
                            searchableCafes = result.data.searchableCafes,
                            pendingClaims = result.data.pendingClaims,
                            isLoading = false
                        )
                    }
                }
                is AppResult.Failure -> {
                    _uiState.update { state ->
                        state.copy(
                            ownedCafes = emptyList(),
                            searchableCafes = emptyList(),
                            pendingClaims = emptyList(),
                            isLoading = false,
                            infoMessage = result.error.toString()
                        )
                    }
                }
            }
        }
    }

    private fun refreshPendingClaims(resetMessage: Boolean = true) {
        viewModelScope.launch {
            when (val result = getCafeManagementUseCase.invoke()) {
                is AppResult.Success -> {
                    _uiState.update { state ->
                        state.copy(
                            ownedCafes = result.data.ownedCafes,
                            searchableCafes = result.data.searchableCafes,
                            pendingClaims = result.data.pendingClaims,
                            infoMessage = if (resetMessage) null else state.infoMessage
                        )
                    }
                }
                is AppResult.Failure -> {
                    _uiState.update { state ->
                        state.copy(
                            infoMessage = if (resetMessage) result.error.toString() else state.infoMessage
                        )
                    }
                }
            }
        }
    }

    private fun clickCafe(cafeId: String) {
        viewModelScope.launch {
            _event.emit(CafeManagementEvent.NavigateToCafeDashboard(cafeId))
        }
    }

    private fun clickCafeDetail(cafeId: String) {
        viewModelScope.launch {
            _event.emit(CafeManagementEvent.NavigateToCafe(cafeId))
        }
    }

    private fun changeCafeSearchQuery(query: String) {
        _uiState.update {
            it.copy(
                cafeSearchQuery = query,
                infoMessage = null
            )
        }
    }

    private fun clickClaimCafe(cafeId: String) {
        viewModelScope.launch {
            when (val result = createCafeOwnerClaimUseCase.invoke(cafeId)) {
                is AppResult.Success -> {
                    refreshPendingClaims(resetMessage = false)
                    _uiState.update {
                        it.copy(infoMessage = MSG_OWNER_CLAIM_REGISTERED)
                    }
                }
                is AppResult.Failure -> {
                    _uiState.update {
                        it.copy(infoMessage = result.error.toString())
                    }
                }
            }
        }
    }

    private fun toggleCafeListExpanded() {
        _uiState.update {
            it.copy(
                isShowingAllCafes = !it.isShowingAllCafes
            )
        }
    }

    private fun clickCreateCafe() {
        viewModelScope.launch { _event.emit(CafeManagementEvent.NavigateToCafeInfoRegistration) }
    }

    private fun dismissInfoMessage() {
        _uiState.update {
            it.copy(
                infoMessage = null
            )
        }
    }

    private fun observeSession() {
        jobs[TaskKey.OBSERVE_SESSION]?.cancel()
        jobs[TaskKey.OBSERVE_SESSION] = viewModelScope.launch {
            observeCurrentUserUseCase.invoke().collectLatest {
                loadCafeManagement()
            }
        }
    }

    private fun observeCafeDetailEvent() {
        jobs[TaskKey.OBSERVE_CAFE_DETAIL_EVENT]?.cancel()
        jobs[TaskKey.OBSERVE_CAFE_DETAIL_EVENT] = viewModelScope.launch {
            cafeDetailEventPublisher.events.collectLatest { event ->
                if (event is CafeDetailEvent.CafeInfoUpdated) {
                    patchCafeInfo(event.cafe)
                }
            }
        }
    }

    private fun observeCafeRegistrationClaimEvent() {
        jobs[TaskKey.OBSERVE_CAFE_REGISTRATION_CLAIM_EVENT]?.cancel()
        jobs[TaskKey.OBSERVE_CAFE_REGISTRATION_CLAIM_EVENT] = viewModelScope.launch {
            cafeRegistrationClaimEventPublisher.events.collectLatest { claimEvent ->
                val shouldRefresh = when (claimEvent) {
                    is CafeRegistrationClaimEvent.Created -> _uiState.value.pendingClaims
                        .any { pendingClaim -> pendingClaim.claimId == claimEvent.claimId }
                    is CafeRegistrationClaimEvent.Approved -> _uiState.value.pendingClaims
                        .any { pendingClaim -> pendingClaim.claimId == claimEvent.claimId }
                    is CafeRegistrationClaimEvent.Rejected -> _uiState.value.pendingClaims
                        .any { pendingClaim -> pendingClaim.claimId == claimEvent.claimId }
                }
                if (shouldRefresh) {
                    refreshPendingClaims(resetMessage = false)
                }
            }
        }
    }

    private fun startClaimPolling() {
        jobs[TaskKey.POLL_CLAIM]?.cancel()
        jobs[TaskKey.POLL_CLAIM] = viewModelScope.launch {
            while (isActive) {
                delay(CAFE_MANAGEMENT_CLAIM_POLLING_INTERVAL_MILLIS)

                if (isActive) {
                    refreshPendingClaims(resetMessage = false)
                }
            }
        }
    }

    private fun patchCafeInfo(cafe: Cafe) {
        _uiState.update { state ->
            state.copy(
                ownedCafes = state.ownedCafes.map { item ->
                    if (item.id == cafe.id) {
                        item.copy(
                            name = cafe.name,
                            city = cafe.region.city,
                            rating = cafe.ratingAvg
                        )
                    } else {
                        item
                    }
                },
                searchableCafes = state.searchableCafes.map { item ->
                    if (item.id == cafe.id) {
                        item.copy(
                            name = cafe.name,
                            location = "${cafe.region.city} ${cafe.region.address}"
                        )
                    } else {
                        item
                    }
                }
            )
        }
    }

    fun onAction(action: CafeManagementAction) {
        when (action) {
            is CafeManagementAction.ClickCafe -> clickCafe(action.cafeId)
            is CafeManagementAction.ClickCafeDetail -> clickCafeDetail(action.cafeId)
            is CafeManagementAction.ChangeCafeSearchQuery -> changeCafeSearchQuery(action.query)
            is CafeManagementAction.ClickClaimCafe -> clickClaimCafe(action.cafeId)
            CafeManagementAction.ToggleCafeListExpanded -> toggleCafeListExpanded()
            CafeManagementAction.ClickCreateCafe -> clickCreateCafe()
            CafeManagementAction.DismissInfoMessage -> dismissInfoMessage()
        }
    }

    init {
        observeSession()
        observeCafeDetailEvent()
        observeCafeRegistrationClaimEvent()
        startClaimPolling()
    }

    override fun onCleared() {
        jobs.values.forEach(Job::cancel)
        jobs.clear()
        super.onCleared()
    }

    private enum class TaskKey {
        LOAD_CAFE_MANAGEMENT,
        OBSERVE_SESSION,
        OBSERVE_CAFE_DETAIL_EVENT,
        OBSERVE_CAFE_REGISTRATION_CLAIM_EVENT,
        POLL_CLAIM
    }

    companion object {
        private const val MSG_OWNER_CLAIM_REGISTERED = "cafemgmt_info_owner_claim_registered"
    }
}

private const val CAFE_MANAGEMENT_CLAIM_POLLING_INTERVAL_MILLIS = 60_000L
